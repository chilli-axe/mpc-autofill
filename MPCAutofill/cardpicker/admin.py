from django.contrib import admin

from .models import Card, DFCPair, Project, ProjectMember, Source, Tag


# Register your models here.
@admin.register(Tag)
class AdminTag(admin.ModelAdmin[Tag]):
    list_display = ("name",)
    search_fields = ("name",)


@admin.register(Card)
class AdminCard(admin.ModelAdmin[Card]):
    list_display = ("identifier", "name", "source", "dpi", "date_created", "tags")
    search_fields = ("identifier", "name")


@admin.register(DFCPair)
class AdminDFCPair(admin.ModelAdmin[DFCPair]):
    list_display = ("front", "back")
    search_fields = ("front",)


@admin.register(Source)
class AdminSource(admin.ModelAdmin[Source]):
    list_display = ("name", "identifier", "contribution", "description")
    search_fields = ("name", "identifier")

    def contribution(self, obj: Source) -> str:
        qty_all, qty_cards, qty_cardbacks, qty_tokens, avgdpi = obj.count()

        return "{} images - {} cards, {} cardbacks, and {} tokens @ {} DPI on average".format(
            qty_all, qty_cards, qty_cardbacks, qty_tokens, avgdpi
        )


@admin.register(Project)
class AdminProject(admin.ModelAdmin[Project]):
    list_display = ("key", "name", "user", "date_created", "date_modified", "cardback", "cardstock")


@admin.register(ProjectMember)
class AdminCardProjectMembership(admin.ModelAdmin[ProjectMember]):
    list_display = ("card_id", "project", "query", "slot", "face")
