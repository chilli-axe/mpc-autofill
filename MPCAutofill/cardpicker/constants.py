import pycountry

DATE_FORMAT = "jS F, Y"
DEFAULT_LANGUAGE = pycountry.languages.get(alpha_2="EN")

NEW_CARDS_PAGE_SIZE = 12
NEW_CARDS_DAYS = 14
EDITOR_SEARCH_MAX_QUERIES = 300
CARDS_PAGE_SIZE = 1000
EXPLORE_SEARCH_MAX_PAGE_SIZE = 100

MAX_SIZE_MB = 30
NSFW = "NSFW"
