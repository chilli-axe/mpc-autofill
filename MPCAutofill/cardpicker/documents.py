from django_elasticsearch_dsl import Document, fields
from django_elasticsearch_dsl.registries import registry
from elasticsearch_dsl import analyzer

from cardpicker.models import Card

# custom elasticsearch analysers are configured here to add the `asciifolding` filter, which handles accents:
# https://www.elastic.co/guide/en/elasticsearch/reference/7.17/analysis-asciifolding-tokenfilter.html
# https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-standard-analyzer.html
precise_analyser = analyzer("precise_analyser", tokenizer="keyword", filter=["apostrophe", "lowercase", "asciifolding"])
fuzzy_analyser = analyzer("fuzzy_analyser", tokenizer="standard", filter=["apostrophe", "lowercase", "asciifolding"])


@registry.register_document
class CardSearch(Document):
    source_pk = fields.TextField(attr="get_source_pk", analyzer="keyword")
    searchq_fuzzy = fields.TextField(attr="searchq", analyzer=fuzzy_analyser)
    searchq_precise = fields.TextField(attr="searchq", analyzer=precise_analyser)
    searchq_keyword = fields.KeywordField(attr="searchq")
    card_type = fields.KeywordField()
    date_created = fields.DateField()
    date_modified = fields.DateField()
    language = fields.TextField(analyzer=precise_analyser)  # case insensitivity is one less thing which can go wrong
    tags = fields.KeywordField()  # all elasticsearch fields support arrays by default

    class Index:
        # name of the elasticsearch index
        name = "cards"
        # see Elasticsearch Indices API reference for available settings
        settings = {"number_of_shards": 5, "number_of_replicas": 0}

    class Django:
        model = Card
        fields = ["identifier", "priority", "dpi", "size"]
