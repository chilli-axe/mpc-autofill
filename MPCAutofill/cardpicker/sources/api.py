import datetime as dt
import functools
import os
import re
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional

import pycountry
import ratelimit
from googleapiclient.discovery import Resource, build
from googleapiclient.errors import HttpError
from oauth2client.service_account import ServiceAccountCredentials

from cardpicker.search import sanitisation
from cardpicker.tags import Tags

thread_local = threading.local()  # Should only be called once per thread


@dataclass
class Folder:
    id: str
    name: str
    parent: Optional["Folder"]

    @functools.cached_property
    def top_level_folder(self) -> "Folder":
        if self.parent is None:
            return self
        return self.parent.top_level_folder

    def get_full_path(self, tags: Tags) -> str:
        _, name, _ = self.unpack_name(tags=tags)
        if self.parent is None:
            return name
        return f"{self.parent.get_full_path(tags=tags)} / {name}"

    def unpack_name(self, tags: Tags) -> tuple[Optional[pycountry.Languages], str, set[str]]:
        """
        The folder's name is unpacked according to the below schema. For example, consider `{EN} Cards [NSFW]`:
             {EN}              Cards         [NSFW]
        └─ language ──┘ └─ folder name ──┘ └─ tags ──┘
        """

        folder_name_results = re.compile(r"^(?:\{(.+)\} )?(.*?)$").search(self.name)
        assert folder_name_results is not None
        language_code, name = folder_name_results.groups()
        language = pycountry.languages.get(alpha_2=language_code) if language_code else None
        name_with_no_tags, extracted_tags = tags.extract_name_and_tags(name)
        return language, sanitisation.fix_whitespace(name_with_no_tags), extracted_tags

    def get_language(self, tags: Tags) -> Optional[pycountry.Languages]:
        language, _, _ = self.unpack_name(tags=tags)
        if self.parent is None:
            return language
        return language if language is not None else self.parent.get_language(tags=tags)

    def get_tags(self, tags: Tags) -> set[str]:
        _, _, extracted_tags = self.unpack_name(tags=tags)
        if self.parent is None:
            return extracted_tags
        return self.parent.get_tags(tags=tags) | extracted_tags


@dataclass
class Image:
    id: str
    name: str
    size: int
    created_time: dt.datetime
    modified_time: dt.datetime
    height: int
    folder: Folder

    def unpack_name(self, tags: Tags) -> tuple[pycountry.Languages, str, set[str], str]:
        """
        The image's name is unpacked according to the below schema. For example, consider `{EN} Opt [NSFW].png`:
             {EN}             opt          [NSFW]   .      png
        └─ language ──┘ └─ card name ──┘ └─ tags ──┘ └─ extension ──┘
        """

        assert self.name, "File name is empty string"
        assert "." in self.name, "File name has no extension"
        extensionless_name, extension = self.name.rsplit(".", 1)
        image_name_results = re.compile(r"^(?:\{(.+)\} )?(.*?)$").search(extensionless_name)
        assert image_name_results is not None
        language_code, name = image_name_results.groups()
        language = pycountry.languages.get(alpha_2=language_code) if language_code else None
        name_with_no_tags, extracted_tags = tags.extract_name_and_tags(name)
        return (
            language or self.folder.get_language(tags=tags),
            sanitisation.fix_whitespace(name_with_no_tags),
            extracted_tags | self.folder.get_tags(tags=tags),
            extension,
        )

    def get_language(self, tags: Tags) -> pycountry.Languages:
        language, _, _, _ = self.unpack_name(tags=tags)
        return language

    def get_tags(self, tags: Tags) -> set[str]:
        _, _, image_tags, _ = self.unpack_name(tags=tags)
        return image_tags


# region google drive API
# Google Drive API usage limits reference: https://developers.google.com/drive/api/guides/limits


# If modifying these scopes, delete the file token.pickle.
SCOPES = ["https://www.googleapis.com/auth/drive.metadata.readonly", "https://www.googleapis.com/auth/drive.readonly"]

SERVICE_ACC_FILENAME = "client_secrets.json"


def find_or_create_google_drive_service() -> Resource:
    if (service := getattr(thread_local, "google_drive_service", None)) is None:
        creds = ServiceAccountCredentials.from_json_keyfile_name(
            str(Path(os.path.abspath(__file__)).parent.parent.parent / SERVICE_ACC_FILENAME), scopes=SCOPES
        )
        service = build("drive", "v3", credentials=creds)
        thread_local.google_drive_service = service
    return service


@ratelimit.sleep_and_retry  # type: ignore  # `ratelimit` does not implement decorator typing correctly
@ratelimit.limits(calls=20_000, period=100)  # type: ignore  # `ratelimit` does not implement decorator typing correctly
def execute_google_drive_api_call(service: Resource) -> Optional[dict[str, Any]]:
    try:
        return service.execute()
    except HttpError:
        return {}


# endregion

__all__ = ["Image", "Folder", "find_or_create_google_drive_service", "execute_google_drive_api_call"]
