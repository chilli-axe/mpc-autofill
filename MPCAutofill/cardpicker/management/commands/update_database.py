import time
from typing import Any, Optional

from django.core.management.base import BaseCommand

from cardpicker.models import Source
from cardpicker.search.search_functions import ping_elasticsearch
from cardpicker.sources.update_database import update_database
from cardpicker.utils import log_hours_minutes_seconds_elapsed


class Command(BaseCommand):
    help = "You may specify one of the following drives: " + ", ".join(Source.objects.values_list("key", flat=True))

    def add_arguments(self, parser) -> None:  # type: ignore
        parser.add_argument("-d", "--drive", type=str, help="Only update a specific drive")

    def handle(self, *args: Any, **kwargs: str) -> None:
        if not ping_elasticsearch():
            raise Exception("Elasticsearch is offline!")
        # user can specify which drive should be searched - if no drive is specified, search all drives
        drive: Optional[str] = kwargs.get("drive", None)
        t0 = time.time()
        update_database(source_key=drive)
        log_hours_minutes_seconds_elapsed(t0)
