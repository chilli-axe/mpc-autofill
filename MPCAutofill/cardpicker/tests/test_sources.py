import datetime as dt

import freezegun
import pytest

from django.core import management
from django.utils.timezone import make_aware, make_naive

from cardpicker.documents import CardSearch
from cardpicker.models import Card
from cardpicker.sources.api import Folder, Image
from cardpicker.sources.update_database import bulk_sync_objects, update_database
from cardpicker.tags import Tags
from cardpicker.tests import factories

DEFAULT_DATE = dt.datetime(2023, 1, 1)


class TestAPI:
    # region constants

    FOLDER_A = Folder(id="a", name="Folder A", parent=None)
    FOLDER_B = Folder(id="b", name="Folder B", parent=FOLDER_A)
    FOLDER_C = Folder(id="c", name="Folder C [NSFW]", parent=FOLDER_B)
    FOLDER_D = Folder(id="d", name="Folder D [Tag in data]", parent=FOLDER_B)
    FOLDER_E = Folder(id="e", name="Folder E [tagindata]", parent=FOLDER_B)  # refers to the tag's alias
    FOLDER_F = Folder(id="f", name="Folder F [tagindata]", parent=FOLDER_B)  # refers to the tag's alias
    FOLDER_G = Folder(id="g", name="Folder G [Tag in Data] (Some more words)", parent=FOLDER_B)
    FOLDER_H = Folder(id="h", name="Folder H [Tag in Data, Some more words]", parent=None)
    FOLDER_X = Folder(id="x", name="Folder X [NSFW, Extended, Full Art]", parent=None)
    FOLDER_Y = Folder(id="y", name="Folder Y [full art, Invalid Tag]", parent=None)
    FOLDER_Z = Folder(id="z", name="Folder z [Full Art", parent=None)
    FOLDER_FRENCH = Folder(id="french", name="{FR} Folder", parent=None)

    IMAGE_A = Image(
        id="a",
        name="Image A.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_B = Image(
        id="b",
        name="Image B [NSFW].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_C = Image(
        id="b",
        name="Image C.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_C,
    )
    IMAGE_D = Image(
        id="b",
        name="Image D [NSFW, full art].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_C,
    )
    IMAGE_E = Image(
        id="e",
        name="Image E [invalid tag.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_F = Image(
        id="F",
        name="Image F [NSFW, tag in data].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_G = Image(
        id="G",
        name="Image G [NSFW] (John Doe).png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_H = Image(
        id="H",
        name="Image H [A, NSFW, B] (John Doe).png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_I = Image(
        id="I",
        name="Image A.I.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_J = Image(
        id="J",
        name="Image J [Child Tag].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_K = Image(
        id="K",
        name="Image K [Grandchild Tag].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_L = Image(
        id="L",
        name="Image L [NSFW].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_D,
    )
    IMAGE_FRENCH = Image(
        id="french",
        name="French.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_FRENCH,
    )
    IMAGE_ENGLISH = Image(
        id="english",
        name="{EN} English.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_FRENCH,
    )
    IMAGE_NSFW = Image(
        id="nsfw",
        name="NSFW [NSFW].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_DOUBLE_NSFW = Image(
        id="double nsfw",
        name="NSFW (NSFW) [NSFW].png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_A,
    )
    IMAGE_IMPLICITLY_FRENCH = Image(
        id="implicitly_french",
        name="Implicitly French.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_FRENCH,
    )
    IMAGE_EXPLICITLY_ENGLISH = Image(
        id="explicitly_english",
        name="{EN} Explicitly English.png",
        size=1,
        created_time=DEFAULT_DATE,
        modified_time=DEFAULT_DATE,
        height=1,
        folder=FOLDER_FRENCH,
    )

    # endregion

    # region tests

    @pytest.mark.parametrize(
        "folder, full_path",
        [(FOLDER_A, "Folder A"), (FOLDER_B, "Folder A / Folder B"), (FOLDER_C, "Folder A / Folder B / Folder C")],
    )
    def test_folder_full_path(self, django_settings, folder, full_path):
        tags = Tags()
        assert folder.get_full_path(tags=tags) == full_path

    @pytest.mark.parametrize(
        "folder, expected_language",
        [
            (FOLDER_A, None),
            (FOLDER_FRENCH, "FR"),
        ],
    )
    def test_folder_language(self, django_settings, folder, expected_language):
        tags = Tags()
        if expected_language is None:
            assert folder.get_language(tags=tags) is None
        else:
            assert folder.get_language(tags=tags).alpha_2.lower() == expected_language.lower()

    @pytest.mark.parametrize(
        "folder, expected_tags",
        [
            (FOLDER_A, set()),
            (FOLDER_B, set()),
            (FOLDER_C, {"NSFW"}),
            (FOLDER_D, {"Tag in Data"}),
            (FOLDER_E, {"Tag in Data"}),
            (FOLDER_X, {"NSFW", "Extended", "Full Art"}),
            (FOLDER_Y, {"Full Art"}),
            (FOLDER_Z, set()),
        ],
    )
    def test_folder_tags(self, django_settings, tag_in_data, extended_tag, full_art_tag, folder, expected_tags):
        tags = Tags()
        assert folder.get_tags(tags=tags) == expected_tags

    @pytest.mark.parametrize(
        "folder, expected_language, expected_name, expected_tags",
        [
            (FOLDER_A, None, "Folder A", set()),
            (FOLDER_B, None, "Folder B", set()),
            (FOLDER_C, None, "Folder C", {"NSFW"}),
            (FOLDER_G, None, "Folder G (Some more words)", {"Tag in Data"}),
            (FOLDER_H, None, "Folder H [Some more words]", {"Tag in Data"}),
        ],
    )
    def test_folder_name(
        self,
        django_settings,
        tag_in_data,
        extended_tag,
        full_art_tag,
        folder,
        expected_language,
        expected_name,
        expected_tags,
    ):
        tags = Tags()
        language, name, extracted_tags = folder.unpack_name(tags=tags)
        if expected_language is None:
            assert language is None
        else:
            assert language.alpha_2.lower() == expected_language.lower()
        assert name == expected_name
        assert extracted_tags == expected_tags

    @pytest.mark.parametrize(
        "image, expected_language",
        [
            (IMAGE_A, None),
            (IMAGE_FRENCH, "FR"),
            (IMAGE_ENGLISH, "EN"),
            (IMAGE_IMPLICITLY_FRENCH, "FR"),
            (IMAGE_EXPLICITLY_ENGLISH, "EN"),  # despite being in a French folder
        ],
    )
    def test_image_language(self, django_settings, image, expected_language):
        tags = Tags()
        if expected_language is None:
            assert image.get_language(tags=tags) is None
        else:
            assert image.get_language(tags=tags).alpha_2.lower() == expected_language.lower()

    @pytest.mark.parametrize(
        "image, expected_tags",
        [
            (IMAGE_A, set()),
            (IMAGE_B, {"NSFW"}),
            (IMAGE_C, {"NSFW"}),
            (IMAGE_D, {"NSFW", "Full Art"}),
            (IMAGE_E, set()),
            (IMAGE_F, {"NSFW", "Tag in Data"}),
            (IMAGE_H, {"NSFW"}),
            (IMAGE_J, {"Child Tag", "Tag in Data"}),  # `Tag in Data` is implied by `Child Tag`
            (IMAGE_K, {"Grandchild Tag", "Child Tag", "Tag in Data"}),  # `Child Tag` is implied by `Grandchild Tag`
        ],
    )
    def test_image_tags(self, django_settings, grandchild_tag, extended_tag, full_art_tag, image, expected_tags):
        tags = Tags()
        assert image.get_tags(tags=tags) == expected_tags

    @pytest.mark.parametrize(
        "image, expected_language, expected_name, expected_tags, expected_extension",
        [
            (IMAGE_A, None, "Image A", set(), "png"),
            (IMAGE_B, None, "Image B", {"NSFW"}, "png"),
            (IMAGE_C, None, "Image C", {"NSFW"}, "png"),  # tag inherited from parent
            (IMAGE_D, None, "Image D", {"NSFW", "Full Art"}, "png"),
            (IMAGE_E, None, "Image E [invalid tag", set(), "png"),
            (IMAGE_F, None, "Image F", {"NSFW", "Tag in Data"}, "png"),
            (IMAGE_G, None, "Image G (John Doe)", {"NSFW"}, "png"),
            (IMAGE_H, None, "Image H [A, B] (John Doe)", {"NSFW"}, "png"),
            (IMAGE_I, None, "Image A.I", set(), "png"),
            (IMAGE_L, None, "Image L", {"NSFW", "Tag in Data"}, "png"),  # first tag from folder, second from image
            (IMAGE_NSFW, None, "NSFW", {"NSFW"}, "png"),
            (IMAGE_DOUBLE_NSFW, None, "NSFW", {"NSFW"}, "png"),
        ],
    )
    def test_unpack_name(
        self,
        django_settings,
        tag_in_data,
        extended_tag,
        full_art_tag,
        image,
        expected_language,
        expected_name,
        expected_tags,
        expected_extension,
    ):
        tags = Tags()
        language, name, extracted_tags, extension = image.unpack_name(tags=tags)
        if expected_language is None:
            assert language is None
        else:
            assert language.alpha_2.lower() == expected_language.lower()
        assert name == expected_name
        assert extracted_tags == expected_tags
        assert extension == expected_extension


# endregion


class TestUpdateDatabase:
    # region tests

    def test_comprehensive_snapshot(self, snapshot, django_settings, elasticsearch, all_sources, tag_in_data):
        update_database()
        assert list(Card.objects.all().order_by("identifier")) == snapshot(name="cards")

    def test_upsert(self, django_settings, elasticsearch, all_sources):
        update_database()
        pk_to_identifier_1 = {x.pk: x.identifier for x in Card.objects.all()}
        update_database()
        pk_to_identifier_2 = {x.pk: x.identifier for x in Card.objects.all()}
        assert pk_to_identifier_1 == pk_to_identifier_2

    @pytest.mark.parametrize(
        "existing_cards, incoming_cards",
        [
            pytest.param(
                [],
                [],
                id="no changes to empty database",
            ),
            pytest.param(
                [("existing", "Existing Card", DEFAULT_DATE, tuple())],
                [("existing", "Existing Card", DEFAULT_DATE, tuple())],
                id="no changes to populated database",
            ),
            pytest.param(
                [],
                [("created", "Created Card", DEFAULT_DATE, tuple())],
                id="create one card",
            ),
            pytest.param(
                [("updated", "Card to Update", DEFAULT_DATE, tuple())],
                [("updated", "Updated Card", DEFAULT_DATE + dt.timedelta(days=1), tuple())],
                id="update one card",
            ),
            pytest.param(
                [("updated", "Card to Update (Tag in Data)", DEFAULT_DATE, tuple())],
                [("updated", "Updated Card (Tag in Data)", DEFAULT_DATE, ("Tag in Data",))],
                id="update one card - changes to tags but not modified on source side",
            ),
            pytest.param(
                [("deleted", "Card to Delete", DEFAULT_DATE, tuple())],
                [],
                id="delete one card",
            ),
            pytest.param(
                [("updated", "Card to Update", DEFAULT_DATE, set()), ("deleted", "Card to Delete", DEFAULT_DATE, [])],
                [
                    ("created", "Created Card", DEFAULT_DATE, tuple()),
                    ("updated", "Updated Card", DEFAULT_DATE + dt.timedelta(days=1), tuple()),
                ],
                id="create + update + delete",
            ),
            pytest.param(
                [("existing", "Existing Card", DEFAULT_DATE, tuple())],
                [
                    ("existing", "Existing Card", DEFAULT_DATE, tuple()),
                    ("created", "Created Card", DEFAULT_DATE, tuple()),
                ],
                id="create one card while another card exists and is not modified",
            ),
        ],
    )
    @freezegun.freeze_time(DEFAULT_DATE)
    def test_bulk_sync_objects(
        self, django_settings, elasticsearch, tag_in_data, example_drive_1, existing_cards, incoming_cards
    ):
        # arrange - set up database and elasticsearch according to `existing_cards`
        source = factories.SourceFactory()
        for (identifier, searchq, date_modified, tags) in existing_cards:
            factories.CardFactory(
                identifier=identifier,
                searchq=searchq,
                date_created=make_aware(DEFAULT_DATE),
                date_modified=make_aware(date_modified),
                source=source,
                tags=list(tags),
            )
        management.call_command("search_index", "--rebuild", "-f")

        # act
        bulk_sync_objects(
            source=source,
            cards=[
                Card(
                    identifier=identifier,
                    searchq=searchq,
                    date_created=make_aware(DEFAULT_DATE),
                    date_modified=make_aware(date_modified),
                    source=source,
                    tags=list(tags),
                    # not strictly relevant for this test, but values for these non-nullable fields are required.
                    size=0,
                )
                for (identifier, searchq, date_modified, tags) in incoming_cards
            ],
        )

        # assert - database and elasticsearch should now match `incoming_cards`
        assert {
            (card.identifier, card.searchq, make_naive(card.date_modified), tuple(sorted(card.tags)))
            for card in Card.objects.all()
        } == set(incoming_cards)
        assert {
            (result.identifier, result.searchq_keyword, make_naive(result.date_modified), tuple(sorted(result.tags)))
            for result in CardSearch().search().scan()
        } == set(incoming_cards)

    # endregion
