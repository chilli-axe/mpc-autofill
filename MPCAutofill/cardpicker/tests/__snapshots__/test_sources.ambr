# name: TestUpdateDatabase.test_comprehensive_snapshot[cards]
  list([
    <Card: [Example Drive 1] <EN> Mountain [Type: CARD, Identifier: 1-dcs0FEE05MTGiYbKqs9HnRdhXkgtIJG, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] <EN> Plains [Type: CARD, Identifier: 1743iShpNEhY5xQ2PCUNpca4UL5t4rG3g, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] <EN> Delver of Secrets [Type: CARD, Identifier: 17fopRCNRge72U8Hac8pApHZtEalx5kHy, Uploaded: 09/10/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Huntmaster of the Fells [Type: CARD, Identifier: 1991MWCur9NdAFi-tQQD5YbQj2oqV_WRy, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Island (William Bradford) [Type: CARD, Identifier: 1HsvTYs1jFGe1c8U1PnNZ9aB8jkAW7KU0, Uploaded: 22/09/2022, Priority: 6]>,
    <Card: [Example Drive 1] <EN> Island [Type: CARD, Identifier: 1IDtqSjJ4Yo45AnNA4SplOiN7ewibifMa, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] <EN> Simple Cube [Type: CARDBACK, Identifier: 1JtXL6Ca9nQkvhwZZRR9ZuKA9_DzsFf1V, Uploaded: 22/09/2022, Priority: 17]>,
    <Card: [Example Drive 1] <EN> Swamp [Type: CARD, Identifier: 1O5Y_lBOBJK1D9Tw6pcsQJXYoTZXhR3mV, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] <EN> Forest [Type: CARD, Identifier: 1Sne_vn7TlohIbsjt933Kd5k2ff6vFS1I, Uploaded: 22/09/2022, Priority: 7]>,
    <Card: [Example Drive 1] <EN> PÃ¡st in Flames [Type: CARD, Identifier: 1UPdh7J7hScg4ZnxSPJ-EeBYHLp2s3Oz1, Uploaded: 24/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Goblin [Type: TOKEN, Identifier: 1V5E0avDmNyEUuFfYwx3nA05aj-1HY0rA, Uploaded: 09/10/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Brainstorm [Type: CARD, Identifier: 1c4M-sK9gd0Xju0NXCPtqeTW_DQTldVU5, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 2] <EN> Past in Flames [Type: CARD, Identifier: 1dxSLHtw-VwwE09pZCA8OA6LbuWRZPEoU, Uploaded: 24/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Ravager of the Fells [Type: CARD, Identifier: 1lv8WC1Xf1qxA7VHSc8jOtT5up6FwaBPH, Uploaded: 22/09/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Insectile Aberration [Type: CARD, Identifier: 1mO73GTYlieP0kiZEkF58pJSrZTmC9lNh, Uploaded: 09/10/2022, Priority: 2]>,
    <Card: [Example Drive 1] <EN> Simple Lotus [Type: CARDBACK, Identifier: 1oigI6wz0zA--pNMuExKTs40kBNH6VRP_, Uploaded: 22/09/2022, Priority: 7]>,
  ])
# ---
